### PR TITLE
KAFKA-15361: Process and persist dir info with broker registration

### PR DESCRIPTION
**Pull request by soarez**
Controllers should process and persist directory information from the broker registration request

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by soarez** [comment]
This builds on #14820  - KAFKA-15886: Always specify directories for new partition registrations – so this is marked as draft until #14820 is merged.

**Reviews**: please focus on the second commit, titled "[KAFKA-15361: Process and persist dir info with broker registration](https://github.com/apache/kafka/pull/14838/commits/aa945193cc705df6e6e904b5ab8a9ecfeca4a38b)"

**Comment by soarez** [comment]
@cmccabe @pprovenzano @rondagostino PTAL

File: server-common/src/main/java/org/apache/kafka/common/DirectoryId.java
Comment (pprovenzano) [inline server-common/src/main/java/org/apache/kafka/common/DirectoryId.java]: We should only allow for a directory ID to be MIGRATING if we are in migration mode.

Comment (soarez) [reply server-common/src/main/java/org/apache/kafka/common/DirectoryId.java]: Can you clarify what you mean with "migration mode"?

If you mean zk->kraft migration, then I don't think so. Metadata records may have been persisted in an older version, without directory assignment. That may happen outside of a ZK->KRaft migration.
When such records are found, the default directory for every replica is assumed to be `MIGRATING`.

See in #14820 :
 * `new PartitionRegistration(PartitionRecord record)`
 * `partitionRegistration.merge(PartitionChangeRecord record)`
 * `partitionRegistration.toRecord(...)`
 * `partitionChangeBuilder.setAssignmentChanges()`

**Comment by soarez** [comment]
@cmccabe @rondagostino @pprovenzano please take a look

File: metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java
Comment (rondagostino) [inline metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java]: Can we rename this to `directoriesSorted` or `sortedDirectories` to indicate that this list must be sorted?  We depend on it being sorted when invoking `DirectoryId.isOnline()`.  And when I look at where we invoke that method, it seems that the field `BrokerRegistration.directories` should get a similar rename.

The other alternative is to remove the constraint that the list be sorted within `DirectoryId.isOnline()`.  Since we don't expect to have a large number of directories, I would lean toward this option in order to avoid bugs down the road if we forget to sort a list somewhere.  Could we then remove the sorting of the list in this class and within `BrokerRegistration` and in any tests?  Or am I missing something and we really do need to keep these sorted?

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (rondagostino) [inline metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: This change seems a bit odd to me.  I think it will work, and in fact I think it will do exactly what was being done before, but you moved the creation of the broker entry in the heartbeat manager's internal map into the `register()` method and removed it from the `touch()` method -- which I think is a good change.  But now this reads weirdly because we aren't registering a broker -- the broker should already have been registered by this point, and we just want to "touch" it.  I think this change can be reverted?

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (rondagostino) [inline metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: Is `DuplicateBrokerRegistrationException` appropriate, do you think?  Seems like `BrokerIdNotRegisteredException` might be better, and then we would have to make an update in `testRegisterWithDuplicateDirectoryId()` to check for that other exception class.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java
Comment (rondagostino) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:350]: Nice catch

File: metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java
Comment (rondagostino) [inline metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:283]: Maybe we want to add in `MetadataVersion.LATEST_PRODUCTION` as well?

**Review by rondagostino — COMMENTED** [review]
Looks pretty good, just a few small things.

File: metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java]: The point was to use binary search for lookups, see `DirectoryId.isOnline`. Since we don't expect a large list of directories, it may be ok to always to a O(N) search. We can come back to this in the future if it proves to be an issue. I'll change it to avoid any sorting.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: This is in `clusterControlManager.activate()`, a couple of lines above we reset the heartbeatManager with `heartbeatManager = new BrokerHeartbeatManager(logContext, time, sessionTimeoutNs);`. So no broker will be registered yet with that fresh manager.

`register()` here means _register the broker in the heartbeat manager_, not _register the broker with the controller quorum_. 

If we revert this line change we'll trigger `throw new IllegalStateException("Broker " + brokerId + " is not registered.");` in `brokerHeartbeatManager.touch()`.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: Good point. `BrokerIdNotRegisteredException` is better indeed.

File: metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java
Comment (soarez) [reply metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:283]: Added

**Comment by rondagostino** [comment]
Looks like there are lots of test failures, too.  Seems like you will have to look (e.g. https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/9/tests/) and figure out what is going on.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (rondagostino) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: Ah, okay, yes, makes sense.  Thanks.

File: metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java]: Can you just make it sorted? The nondeterminism of unsorted just sucks for testing / debugging / etc.

File: server-common/src/main/java/org/apache/kafka/common/DirectoryId.java
Comment (cmccabe) [reply server-common/src/main/java/org/apache/kafka/common/DirectoryId.java]: > If you mean zk->kraft migration, then I don't think so. Metadata records may have been persisted in an older version, without directory assignment

yes, agreed. @pprovenzano , `MIGRATING` state can happen outside ZK migration.

Honestly, I think `MIGRATING` isn't a great name. Someone who sticks on metadata version 15 won't ever "migrate" to having directory IDs. Maybe calling this `UNKNOWN` would be more descriptive? But that's another discussion I guess :)

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: I don't see why we need to handle directories in `HeartbeatManager`. I made this comment in the other PR too. HBM is about timing out brokers. It's not about JBOD stuff really.

Directories are hard state so they should be kept in `ClusterControlManager`. And we need to issue a new `BrokerRegistrationRecord` when they change.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: Hmm. `DuplicateBrokerRegistrationException` is certainly the wrong thing, but I don't think `BrokerIdNotRegisteredException` is quite right either. That exception was intended for when you made an RPC that required broker N to be registered (like heartbeat) but broker N was not in fact registered. It wasn't intended for an error returned from the registration RPC itself. After all, it provides no information. OK the broker is not registered, but why?

I can see that there are some lines above which are using `BrokerIdNotRegisteredException` as a generic "didn't register" error, but that doesn't seem quite right...
```
        if (request.isMigratingZkBroker() && !zkRegistrationAllowed()) {
            throw new BrokerIdNotRegisteredException("Controller does not support registering ZK brokers.");
        }

        if (!request.isMigratingZkBroker() && featureControl.inPreMigrationMode()) {
            throw new BrokerIdNotRegisteredException("Controller is in pre-migration mode and cannot register KRaft " +
                "brokers until the metadata migration is complete.");
        }
```

The "pre-migration" thing is just a straightforward logic error that should never happen. Maybe `RuntimeException` (and consequent controller failover) is OK for that.

The `request.isMigratingZkBroker()` check probably deserves its own error code since it's a weird scenario that should have been given its own error code by the migration KIP. Not sure if it's worth doing now but... probably

Your thing is more like INVALID_REQUEST, since someone is sending obvious nonsense which superficially resembles the schema.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:350]: Sigh. Looks like this was introduced in KAFKA-14538, which I reviewed. It really only bites in scenarios where you're running a pre-NO_OP record MV, which is rare these days.

Thanks for the fix.

**Review by pprovenzano — APPROVED** [review]
LGTM

File: server-common/src/main/java/org/apache/kafka/common/DirectoryId.java
Comment (soarez) [reply server-common/src/main/java/org/apache/kafka/common/DirectoryId.java]: I don't think it's a bad name, we just need to make it clear the migration refers to the record version, and not to the cluster migration. 

You might not remember, but [the name was your suggestion](https://lists.apache.org/thread/38sbj42ssoksjq2p488vdk1992dw3j1t). 😄

File: metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java]: Reverted, back to a sorted list, but now clearly named as such.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: > And we need to issue a new BrokerRegistrationRecord when they change.
We do that in #14902 which builds on this.

I have removed directory state from the heartbeat manager, we now rely only on `ClusterControlManager`, and any persisted metadata state.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: Updated all three cases to throw `InvalidRequestException` instead.

File: server-common/src/main/java/org/apache/kafka/common/DirectoryId.java
Comment (cmccabe) [reply server-common/src/main/java/org/apache/kafka/common/DirectoryId.java]: Heh. :)

Yeah I'm on the fence. With good enough docs maybe the name is ok...

**Review by rondagostino — APPROVED** [review]
LGTM assuming tests pass.  Thanks for the PR and adjustments along the way!

**Comment by soarez** [comment]
Thanks for the review.

@rondagostino AFAICT, the tests are passing on all JDK versions. Did I miss something?

**Comment by rondagostino** [comment]
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/13/tests
```
208 tests have failed
There are 20 new tests failing, 188 existing failing and 272 skipped
```
@soarez it seems that something is not correct.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (cmccabe) [inline metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:367]: It should not be legal to send an empty `logDirs` array when metadataVersion >= 3.7-IV2

Since setting that MV on the controller requires that all the brokers support it.

It is, of course, OK to send logDirs when MV < 3.7-IV2. They will be ignored

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (cmccabe) [inline metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: I think we should have a custom error code for this situation. InvalidRequestException seems too generic.

Also, we shouldn't be checking the previous registration for this specific broker ID... it would be quite normal to re-register broker ID X with the same set of directories it had last time we registered it.

File: metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java
Comment (cmccabe) [inline metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:195]: Seems like we should copy the list and sort it here to make sure that it actually is sorted. It should be a short list (a dozen or so max) so not a big performance issue.

File: metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java
Comment (cmccabe) [inline metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:344]: This is another reason to go with sorted -- `list.equals` won't do the right thing otherwise.

File: metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java
Comment (cmccabe) [inline metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java]: I feel like we should explicitly include `IBP_3_7_IV2` here rather than `LATEST_PRODUCTION`. After all, that is the first JBOD version so it's worth testing specifically, even after `LATEST_PRODUCTION` marches on.

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: Another thing. This is `O(num_brokers * num_directories)` which isn't good for larger deployments. I think we should just have a `TimelineHashMap<Uuid, Integer>` that maps log dir UUID to the broker that has it. The main hassle is removing the mappings during broker unregistration, but that is rare, of course.

File: metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java
Comment (rondagostino) [reply metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:195]: We do that in the prior 2 lines

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: > Also, we shouldn't be checking the previous registration for this specific broker ID... it would be quite normal to re-register broker ID X with the same set of directories it had last time we registered it

This might be what's causing the test failures btw

**Comment by cmccabe** [comment]
Sorry, one more thing: I didn't understand the purpose of the HeartbeatManager changes (now that we've agreed to keep Directory out of HBM). Seems like we should be able to keep this file the same now? Or if not, can we change it in a separate PR where we have some rationale?

File: metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:195]: ah, sorry, that is my bad. Please ignore this then :)

File: metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java
Comment (soarez) [reply metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java]: @cmccabe:

>  InvalidRequestException seems too generic

I've introduced InvalidRegistrationException to handle these.

> Also, we shouldn't be checking the previous registration for this specific broker ID.

I forgot that. Thanks for pointing this out!

> we should just have a TimelineHashMap<Uuid, Integer> that maps log dir UUID to the broker that has it

Great suggestion. Added.

File: metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java
Comment (soarez) [reply metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java]: Replaced `LATEST_PRODUCTION` with `IBP_3_7_IV2`.

**Comment by soarez** [comment]
@cmccabe:

> I didn't understand the purpose of the HeartbeatManager changes (now that we've agreed to keep Directory out of HBM). Seems like we should be able to keep this file the same now? Or if not, can we change it in a separate PR where we have some rationale?

The changes aren't necessary, but they are an improvement. I've moved them to a separate PR: #14934

**Comment by rondagostino** [comment]
[Still 110 failing tests](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/14/tests), which seems like a lot, especially since https://github.com/apache/kafka/pull/14934/ only shows [18 failing tests](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14934/1/tests/).

I've [restarted the build](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/15/pipeline) in the hope that we see many fewer test failures.

**Comment by soarez** [comment]
@rondagostino I'm trying to figure the test failures out, I've not yet been able to reproduce any failures locally.

**Comment by rondagostino** [comment]
[Latest builds for JDKs 8 and 21 failed](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/16/pipeline/).  Restarted the build: https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/17/pipeline

So frustrating.

**Comment by rondagostino** [comment]
I think the issue is not with our PR.  Here's another [PR that is basically a 2-line change in a single test](https://github.com/apache/kafka/pull/14944) and that has [144 test failures!](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14944/2/tests).

**Comment by rondagostino** [comment]
Looking at the [trunk build jobs](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/activity?branch=trunk):

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2432/tests completed 6 days ago with `25 tests have failed.  There are 21 new tests failing, 4 existing failing `

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2437/tests completed 5 days ago with `60 tests have failed.  There are 36 new tests failing, 24 existing failing`

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2442/tests completed 3 days ago with `30 tests have failed.  There are 14 new tests failing, 16 existing failing`

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2444/tests completed 2 days ago with `53 tests have failed.  There are 37 new tests failing, 16 existing failing`

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2445/tests completed 2 days ago with `92 tests have failed.  There are 17 new tests failing, 75 existing failing`

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2452/tests completed 11 hours ago with `111 tests have failed.  There are 49 new tests failing, 62 existing failing`

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2453/tests completed 7 hours ago with `24 tests have failed.  There are 24 new tests failing, 0 existing failing`

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2454/tests completed 1 hour ago with `101 tests have failed.  There are 39 new tests failing, 62 existing failing`

It feels to me like this is just flakiness -- either in the test or in the infrastructure.  There's no rhyme or reason to it.  I am inclined to commit this PR.  WDYT @cmccabe?

**Comment by cmccabe** [comment]
@rondagostino : I think we can merge this now because I don't think it's adding more failures. I do think we need to do another sweep looking for people calling `Exit.exit`, but that doesn't need to block here.

**Comment by lucasbru** [comment]
There are some tests that seem to have gotten a lot worse since this PR was merged, as far as I can tell, e.g. `ZkMigrationIntegrationTest`.

https://ge.apache.org/scans/tests?search.rootProjectNames=kafka&search.timeZoneId=Europe%2FBerlin&tests.container=kafka.zk.ZkMigrationIntegrationTest&tests.test=testMigrateTopicDeletions(ClusterInstance)%5B6%5D

**Comment by dajac** [comment]
@soarez Could you take a look?

**Comment by soarez** [comment]
@dajac: I'm having a look. `ZkMigrationIntegrationTest.testMigrateTopicDeletions` and `ZkMigrationIntegrationTest.testDualWrite` are consistently failing for `IBP_3_7_IV2` or greater, suggesting this is indeed JBOD related.

**Comment by soarez** [comment]
@dajac, @lucasbru: Please take a look at #14976

**Pull request merged**
